From accb9f7cbfddca6ff277fe110fbc3f506f0e910c Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 12 Jan 2021 21:45:04 +0200
Subject: [PATCH] [SW-20719] string: add memset32() function

Add memset32() function (copied from Linux kernel) for setting memory
areas with 4 bytes alignment.
This is in preparation for clearing SRAM code function addition.

Change-Id: I052abbd1c6c3ca1ec366eb2351d113c76ee08f5e
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/string.c b/common/string.c
index c213d17..4b3b55d 100644
--- a/common/string.c
+++ b/common/string.c
@@ -6,6 +6,7 @@
  */
 
 #include <stddef.h>
+#include <stdint.h>
 #include <string.h>
 
 /**
@@ -123,6 +124,25 @@
 }
 
 /**
+ * memset32() - Fill a memory area with a uint32_t
+ * @s: Pointer to the start of the area.
+ * @v: The value to fill the area with
+ * @count: The number of values to store
+ *
+ * Differs from memset() in that it fills with a uint32_t instead
+ * of a byte.  Remember that @count is the number of uint32_ts to
+ * store, not the number of bytes.
+ */
+void *memset32(uint32_t *s, uint32_t v, size_t count)
+{
+	uint32_t *xs = s;
+
+	while (count--)
+		*xs++ = v;
+	return s;
+}
+
+/**
  * strcmp - Compare two strings
  * @cs: One string
  * @ct: Another string
diff --git a/include/string.h b/include/string.h
index aa0f67f..6c97643 100644
--- a/include/string.h
+++ b/include/string.h
@@ -8,6 +8,7 @@
 #define _PREBOOT_STRING_H_
 
 #include <stddef.h>
+#include <stdint.h>
 
 typedef unsigned long		ulong;
 
@@ -15,6 +16,7 @@
 void *memcpy(void *dest, const void *src, size_t count);
 void *memmove(void * dest,const void *src,size_t count);
 void *memset(void * s,int c,size_t count);
+void *memset32(uint32_t *s, uint32_t v, size_t count);
 
 int memcmp(const void * cs,const void * ct, size_t count);
 int strcmp(const char * cs,const char * ct);
